### PR TITLE
chore: increase async rpc performance

### PR DESCRIPTION
**Pull request by daniel-sanche**
While developing the bigtable client, we found that  async gapic clients are very slow compared to raw grpc calls. Looking at the code, I found a couple low-hanging-fruit optimizations, some just needing a port over from the synchronous client code:
- use cached wrapped functions, instead of wrapping the underlying grpc call before each rpc call
- check if the input is the expected proto type, and avoid creating a new copy if so

These changes give ~5x speed up on a quick benchmark I threw together (1.99s -> 0.428s for unary, 1.839s -> 0.335s for streams. Network mocked out. 10,000 rpc tests), and bring rpc performance more in-line with the sync clients

### CODE REVIEW

**Review by parthea — COMMENTED** [review]
Since this fixes a performance issue, please can you add a test to avoid regressions in peformance?

**Comment by daniel-sanche** [comment]
Ok, I added some tests that make sure that wrap_method is cached instead of being called on each rpc call. Let me know if that works

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (parthea) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2:151]: nit: consolidate these 2 if statements

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (parthea) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2:1714]: nit: revert

**Review by parthea — APPROVED** [review]

**Comment by parthea** [comment]
LGTM, but please wait for review from @vchudnov-g

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (vchudnov-g) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2:157]: Could we make this stanza conditional on the method in fact being an Operation method?

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (vchudnov-g) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2:1186]: Could we make this stanza conditional on the method in fact being an Operation method?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2:385]:
```suggestion
        """ Precompute the wrapped methods, overriding the base class method to use async wrappers."""
```

**Review by vchudnov-g — APPROVED** [review]
Looks good. Thanks for doing this.

I left some minor comments and nits, the most important being making the test stanza for Opeeration methods being conditional.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2
Comment (daniel-sanche) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:332]: Sure, but keep in mind this was [taken directly from the sync version](https://github.com/googleapis/gapic-generator-python/blob/e403acc9a209a5c74f54bbdc8dfff95cea53eaee/gapic/templates/%25namespace/%25name_%25version/%25sub/services/%25service/_client_macros.j2#L107). Should we update this there too?

Comment (daniel-sanche) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:332]: I went ahead and changed the sync version to match. Let me know if that works

File: tests/integration/goldens/asset/tests/unit/gapic/asset_v1/test_asset_service.py
Comment (daniel-sanche) [reply tests/integration/goldens/asset/tests/unit/gapic/asset_v1/test_asset_service.py]: I tried using `if method.lro`, but it looks like a fragment test is still failing. Any tips?

**Pull request merged**
